feat($compile): add one-way binding to the isolate scope defintion

Narretz opened the pull request.

Narretz committed:
feat($compile): add one-way binding to the isolate scope defintion

gkalpak commented on src/ng/compile.js:
```diff
@@ -3032,6 +3052,29 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+            }
+
+            if (definition.collection) {
+              removeWatch = scope.$watchCollection($parse(attrs[attrName]), onParentValueChange);
```
Couldn't `parentGet` be used here (to avoid double parsing) ?

Narretz replied:
Do we even need to parse this? We are just passing the string to watch, which parses this anyway. Yeah, I think I was confused about what's going oon in the two-way binding

thorn0 replied:
I think we should parse it to init the child value. However, you can do this
```js
scope.$watchCollection(parentGet, ...)
```
to not parse twice.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3533,6 +3533,14 @@ describe('$compile', function() {
             colrefAlias: '=* colref',
+            owRef: '<',
+            owRefAlias: '< owRef',
+            owReference: '<',
```
Why both `owRef` and `owReference` ?

gkalpak commented on src/ng/compile.js:
```diff
@@ -186,6 +186,27 @@
+ *     object in ther parent scope with a new one, otherwise the one-way binding won't pick up changes.
+ *
+ *   One-way binding is useful if you do not plan to propagate changes to your isolated scope bindings
+ *   back to the parent. It does not however not make this completely impossible.
```
_not ... not_  -->  _not ..._

thorn0 replied:
I'm reading and re-reading the Caveat No 2 and can't get it. Why do I have to replace something in the parent scope? Isolated components don't care about and don't touch the parent scope.

Narretz replied:
Hehe, yeah the explanation was problematic to write. Basically, if you replace the bound object in the isolate scope, changing only a property of the object in the parent scope won't update the child scope, because the watch on the parent scope only picks up when the identity of the parent scope value changes.

thorn0 replied:
BTW, please have a look at #13921. The paragraph about `=`-bindings, which you took as a starting point, is pretty outdated and inaccurate.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4227,299 @@ describe('$compile', function() {
+      it('should update isolate when the identity of origin changes', inject(function() {
+        compile('<div><span my-component ow-ref="obj">');
+        expect(componentScope.owRef).toBe(undefined);
+        expect(componentScope.owRefAlias).toBe(componentScope.owRef);
```
This only checks that `componentScope.owRefAlias` is `undefined`.
Shouldn't there be similar expectations below ?

Narretz replied:
Updated

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4227,299 @@ describe('$compile', function() {
 
 
+    describe('one-way binding', function() {
+      it('should update isolate when the identity of origin changes', inject(function() {
```
Unecessary `inject()` (here and below) ?

Narretz replied:
They are necessary. (many other tests have them). I  think it's because the $rootScope etc. "globals" are set up in a module'S return fn.

thorn0 commented on src/ng/compile.js:
```diff
@@ -3032,6 +3052,29 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+                destination[scopeName] = newParentValue;
+              });
+            } else {
+              removeWatch = scope.$watch(parentGet), function onParentCollectionValueChange(newParentValue) {
```
The names of the watch listener functions should switch places.

campersau commented on src/ng/compile.js:
```diff
@@ -3032,6 +3052,29 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+                destination[scopeName] = newParentValue;
+              });
+            } else {
+              removeWatch = scope.$watch(attrs[attrName], function onParentCollectionValueChange(newParentValue) {
```
Are the function names for the watch callbacks swapped?
Should't it be `onParentValueChange` here and `onParentCollectionValueChange` above?

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -186,6 +186,27 @@
  *   you want to shallow watch for changes (i.e. $watchCollection instead of $watch) you can use
  *   `=*` or `=*attr` (`=*?` or `=*?attr` if the property is optional).
  *
+ * * `<` or `<attr` - set up one-way (one-directional) binding between a local scope property and the
```
I think the character should be `>` giving more feeling of forward movement

thorn0 replied:
As I wrote in #13854:

> `bindings: { propertyName: "<attrName" }`
> Which means: the value is copied from the attribute `attrName` to the property `propertyName`.

petebacondarwin replied:
Trouble is that in my mind `<attrName` looks like a broken HTML element

petebacondarwin replied:
How about `<=`? It has the same connatation of moving from attribute to property (even more so) it has an element of `=` about it and looks nice `propertyName: '<=attrName'`

thorn0 replied:
Having this connotation is not as important as **not having** the opposite. As for multi-character sigils, I think they would look too scary in combination with `?` and `*`:
```js
bindings: { propertyName: "<=*?attrName" } // oh...
```

petebacondarwin replied:
I'm happy to go with the consensus and get this in :-)

robstove replied:
What was the reason `-` didn't make sense?

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -3533,6 +3533,13 @@ describe('$compile', function() {
+            owRefAlias: '< owRef',
+            owOptref: '<?',
+            owOptrefAlias: '<? owOptref',
+            owOptreference: '<?',
```
OOC, what is the purpose of `owOptref` + `owOptreference` ?

Narretz replied:
Came from a c&p'd test, but `owOptreference` doesn't look like it's needed.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4226,303 @@ describe('$compile', function() {
+      }));
+
+
+      it('should not throw on non assignable changes', inject(function() {
```
Did you mean "non assignable expressions" ?

Narretz replied:
Probably

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4226,303 @@ describe('$compile', function() {
+        var isolateScope = element.isolateScope();
+        expect(isolateScope.owRef).toBeNaN();
+
+        $rootScope.$apply(function(scope) { scope.num = 64; });
```
People don't seem to prefer it, but many `assignment + $apply()` or `$apply(callback)` can be shortened as `$apply('expresion')`. E.g.: `$rootScope.$apply('num = 64')`. Just saying... (it doesn't make any difference anyway :smiley:)

Narretz replied:
This is actually the only test in this block that uses this, because it was added later, all the others use

    $rootScope.num = 64;.
    $rootScope.$apply();

gkalpak replied:
I mean

```js
$rootScope.num = 64;
$rootScope.$apply();
```

can be written much more conviently as

```js
$rootScope.$apply('num = 64');
```

But for some reason, everyone seems to avoid it :smiley:
Really, I don't know why I keep bringing this up, since it doesn't make any difference (in tests).

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4226,303 @@ describe('$compile', function() {
+          componentScope.$apply();
+          expect(componentScope.owRef).toEqual({name: 'b'});
+          expect($rootScope.name).toBe('a');
+        }));
```
I think it would be worth having a test for what happens after changing a property on `$rootScope` after `componentScope.owRef = {name: 'b'}`.
If I am not mistaken, there is no way to overwrite `componentScope.owRef` when using a literal object, right ?

Narretz replied:
I believe so, yes. That's the same with two-way data-binding, the literal always wins.

Narretz replied:
Actually, the test shows currently shows that it is possible.

thorn0 replied:
It's possible even with `=`-bindings now. If a literal is passed, they work like they're one-way.

Narretz replied:
Ah, I overlooked a small detail. Yes, I can add an expectation for what happens when I change name to 'c' on the origin scope (Spoilers, it works)

Narretz replied:
Done

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4226,303 @@ describe('$compile', function() {
+          test('1', 1);
+          test('null', null);
+          test('undefined', undefined);
+          test("'someString'", 'someString');
```
What about `true/false` ?

Narretz replied:
Can do

Narretz committed:
fixes

Narretz committed:
update test expectation

thorn0 commented:
Currently, if an object literal or an array literal is passed as the expression for a two-way (`=`) binding, the set up watch is deep (uses `angular.equals`). The implementation in this PR doesn't do this with the new one-way bindings. IMO it'd more consistent if one-way bindings behaved exactly like two-way bindings, just without probability for the parent to get modified by them (with the known caveats).

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4223,308 @@ describe('$compile', function() {
+
+          $rootScope.name = 'c';
+          $rootScope.$apply();
+          expect(componentScope.owRef).toEqual({name: 'c'});
```
I expected it to remain `{name: 'b'}`. Interesting...

gkalpak replied:
Aha, parsed expression with `inputs` and `$$watchDelegates`. Nice :smiley:

Narretz replied:
That's why it works? I'm actually confused by this right now.

gkalpak replied:
Yes. Parsing `{name: name}` the resulting parsed expression has `inputs`, so it has a `$$watchDelegate` that will re-evaluate the expression whenever the inputs change value (in this case, `inputs` is the `name` property on `$rootScope`).
So, whenever `$rootScope.name` changes, the expression will be evaluated and will return a new object, thus the change will be propagated to the child.

Narretz replied:
So this works, but when you are going deeper, you need objectEquality watch - I've added that right now.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4223,308 @@ describe('$compile', function() {
+        describe('optional one-way binding', function() {
+          it('should update local when origin changes', inject(function() {
+            compile('<div><span my-component ow-optref="name">');
+            expect(componentScope.owOptRef).toBe(undefined);
```
_owOptRef_ --> _owOptref_

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4219,6 +4223,308 @@ describe('$compile', function() {
+          it('should update local when origin changes', inject(function() {
+            compile('<div><span my-component ow-optref="name">');
+            expect(componentScope.owOptRef).toBe(undefined);
+            expect(componentScope.owOptRefAlias).toBe(componentScope.owOptRef);
```
_owOptRefAlias_  -->  _owOptrefAlias_
_owOptRef_  -->  _owOptref_

gkalpak commented on src/ng/compile.js:
```diff
@@ -3032,6 +3053,29 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+            destination[scopeName] = parentGet(scope);
+
+            if (definition.collection) {
+              removeWatch = scope.$watchCollection(attrs[attrName], function onParentCollectionValueChange(newParentValue) {
```
I think we can reuse `parentGet` (here and below) to avoid re-parsing the expression.

Narretz replied:
It works, but I wonder if $parse isn't doing extra work? Or does it detect an already parsed expression?

Narretz commented:
Hm, something strange is going on. The watchAction functions fire twice every time there's an apply. I don't think this should happen. Maybe that's also why the literals work even though they shouldn't.

gkalpak commented:
A couple of minor comments, but LGTM otherwise, expect for the fact that we should be more detailed on what is updated when and what is the difference between `*` and no `*`.

Narretz committed:
deep-watch object literals, fix typos

lgalfaso commented on src/ng/compile.js:
```diff
@@ -3032,6 +3053,29 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+            destination[scopeName] = parentGet(scope);
+
+            if (definition.collection) {
+              removeWatch = scope.$watchCollection(parentGet, function parentCollectionValueWatchAction(newParentValue) {
```
I do not understand what is the use case to use `$watchCollection`. We are not copying things from one side to the other, so I would just drop support for `'<*'`, have the minimal viable feature and cut this release.

If, later on, people ask for this feature, then we can take the time to think if this is the way that best fits the needs.

gkalpak replied:
I agree with @lgalfaso.

The only difference with `<*` is when the parent will be "synced" to the child and this only matters, if the child has set its property to a new instance. Since I can't think of any good usecase for setting the child to a new instance (knowing that at any point the parent could overwrite that value), I believe that `<*` will only cause confusion.

Keeping this feature as simple as possible (which means simpler semantics, less corner-cases, less unexpected behaviors) is pretty important imo.

As @lgalfaso said, we can always add stuff/features, but removing isn't that simple. Let's start simple.
<sub>(BTW, as @lgalfaso pointed out, there is no `*` in ng2, so I think it helps to be consistent.)</sub>

thorn0 replied:
@lgalfaso Well spotted!

jbedard replied:
I just found a use case for this today while using a 2-way binding (and wishing it was a 1-way binding)... when doing things like `<element binding="array | orderBy:'foo'">` using a `=` binding causes an infinite digest while using `=*` works. But `=*` or `<*` really doesn't seem like a great solution for that, it would be great if there was some other way. Maybe things like `filter` and `orderby` should somehow return the same instance each time if it hasn't changed?

lgalfaso replied:
@jbedard when you have `<element binding="array | orderBy:'foo'">` we are not watching the result of the array after it went thru the filter, but before. Anyhow, if the behavior of `filter` and `orderBy` should change, then that should be another issue's discussion

jbedard replied:
> we are not watching the result of the array after it went thru the filter, but before

I don't think that's right. Just doing `$rootScope.$watch("a | orderBy"); $rootScope.$apply("a = []")` causes an infinite digest.  I believe `=*` was created specifically to avoid this. However I think `=*` is a bad solution, it would be better if $parse/$watch handled this somehow...

Narretz committed:
remove shallow watch, update docs

Narretz commented:
I've 
- removed the collectionWatch functionality
- updated the description to incorporate changes made by @thorn0 
- added watching literals by object equality

Narretz committed:
fix typos, wordings

gkalpak commented on src/ng/compile.js:
```diff
@@ -186,6 +186,30 @@
+ *   parent scope. If no `attr` name is specified then the attribute name is assumed to be the same as the
+ *   local name. You can also make the binding optional by adding `?`: `<?` or `<?attr`.
+ *
+ *   For example, given `<dir my-attr="parentModel">` and directive definition of
```
_dir_  -->  _my-component_ (to be consistent with #13921)

gkalpak commented on src/ng/compile.js:
```diff
@@ -186,6 +186,30 @@
+ *     sets the same value. That means if your bound value is an object, changes to its properties
+ *     in the isolated scope will be reflected in the parent scope (because both reference the same object).
+ *     2. one-way binding watches changes to the **identity** of the parent value. That means the
+ *     {@link ng.$rootScope.Scope#$watch`$watch`} on the parent value only fires if the reference
```
Missing space before `` `$watch` ``.

gkalpak commented on src/ng/compile.js:
```diff
@@ -186,6 +186,30 @@
+ *     to know if you one-way bind to an object, and then replace that object in the isolated scope.
+ *     If you now change a property of the object in
+ *     your parent scope, the change will not be propagated to the isolated scope, because the identity
+ *     of the object has not changed. Instead you must assign a new object.
```
I would change _"the identity of the object has not changed"_ to _"the identity of the object in he parent scope has not changed"_.

gkalpak commented:
LGTM (expect for some minor docs typos/fixes).
After this gets merged, it might be a goood idea to add an example in the Directives guide.
